### PR TITLE
feat: add cfg_not_test lint

### LINKED ISSUE #11234: `cfg_not_test`
### What it does

Disallows usage of any conditional compilation that always excludes code from test builds.

### Advantage

- Prevents code being excluded from tests builds. Preventing the case where a codebase appears to have 100% coverage while having untested code.
- Enforces simplicity in testing, guarding against excessive mocking anti-pattern.

### Drawbacks

- May prevent rare legitimate use cases for excluding code from test runs.

### Example

The following would trigger this lint:

```rust
#[cfg(not(test))]
```

```rust
#[cfg(all(debug_assertions, not(test)))]
```

```rust
#[cfg(not(any(not(debug_assertions), test)))]
```

### PR DESCRIPTION
**Pull request by mrnossiom**
Fixes #11234 

changelog: new lint: [`cfg_not_test`]

I don't know whether to lint only the `attr` or also the item associated to it. I guess this would mean putting the check in another place than `check_attribute` but I can't find a way to get the associated item to the attribute.

Also, I'm not sure how to document this lint, I feel like my explications are bad.

### CODE REVIEW

File: clippy_lints/src/cfg_not_test.rs
Comment (Centri3) [inline clippy_lints/src/cfg_not_test.rs:25]:
```suggestion
    /// ### What it does
    /// Checks for usage of `cfg` that excludes code from `test` builds. (i.e., `#{cfg(not(test))]`)
    ///
    /// ### Why is this bad?
    /// This may give the false impression that a codebase has 100% coverage, yet actually has untested code.
    ///
    /// ### Example
    /// ```rust
    /// #[cfg(not(test))]
    /// important_check(); // I'm not actually tested, but not including me will falsely increase coverage!
    /// ```
    /// Use instead:
    /// ```rust
    /// important_check();
    /// ```
```

File: clippy_lints/src/cfg_not_test.rs
Comment (Centri3) [inline clippy_lints/src/cfg_not_test.rs:38]: This won't detect it if `--test` is passed.

This could be a pre-expansion pass lint instead so it's ran before `cfg` is expanded, *but these are very buggy and soft-deprecated*. See #6610 for an example of one of these issues. If it's made into a pre-expansion pass, this should probably be moved to `attrs.rs` and combined with `maybe_misused_cfg`, that way we can minimize its usage to only a few lints.

Whether or not it not detecting it if `--test` is passed is up to you, it's probably not as big of a deal as other lints.

File: clippy_lints/src/cfg_not_test.rs
Comment (Centri3) [inline clippy_lints/src/cfg_not_test.rs:46]: nit: Imo this should give a note instead, like "this will increase code coverage despite not actually being tested"

File: clippy_lints/src/cfg_not_test.rs
Comment (Centri3) [inline clippy_lints/src/cfg_not_test.rs:52]: I believe this can use `has_name` on the attribute instead?

File: clippy_lints/src/cfg_not_test.rs
Comment (MrNossiom) [reply clippy_lints/src/cfg_not_test.rs:46]: maybe but code coverage is not the only issue
(or maybe you meant to add another note, I don't know how to do that... another function, using the `Diasgnostic` struct for more complex lints?)

File: clippy_lints/src/cfg_not_test.rs
Comment (MrNossiom) [reply clippy_lints/src/cfg_not_test.rs:52]: I could have done that, but I think this doesn't check when the test is deeply nested like in the example `#[cfg(not(any(not(debug_assertions), test)))]`

File: clippy_lints/src/cfg_not_test.rs
Comment (MrNossiom) [reply clippy_lints/src/cfg_not_test.rs:46]: `span_lint_and_then()`

File: tests/ui/cfg_not_test.rs
Comment (MrNossiom) [reply tests/ui/cfg_not_test.rs:1]: Why souldn't we lint those ?

File: clippy_lints/src/cfg_not_test.rs
Comment (Centri3) [reply clippy_lints/src/cfg_not_test.rs]: > I didn't know about `cfg` expansion, does this only concern `cfg_attr`?

No, it concerns `cfg` as well. In the HIR, and even AST, `cfg`s are expanded to either *nothing* or unchanged depending on whether it should "be in that compilation run". So, if `not(test)` is successful, it'll be entirely gone before linting occurs. Thus, if the user passes `--test`, there will be no warnings.

For something like `not(test)`, not a big deal, but something like `cfg(all(windows, not(test))` is not good.

I'd recommend messing around a bit on godbolt if you're still having trouble understanding: https://godbolt.org/z/WxPWeM416, you can see the AST before macro-expansion, after, and the HIR by passing different things to `unpretty`.

> (btw, why isn't there more folders for lints?)

Only because nobody's taken the time to :D I'd be 100% in favor of merging a PR to clean that up a bit.

> I don't know if grouping lints has an advantage

It's cleaner, but more importantly, it puts them in the same lint pass which has a *huge* performance boost. For on-by-default (warn/deny) lints, that you almost never want to globally disallow, this is great. In the case of this lint, it's allow-by-default, but I'd still say it's ok in the same lint pass as it does the exact same thing as other lints.

> Also, this deserves his own lint, and IMHO has not much to do with `maybe_misused_cfg`.

I actually agree now; that one is warn-by-default and this one is a restriction lint. So, no complaints from me.

File: tests/ui/cfg_not_test.rs
Comment (Centri3) [reply tests/ui/cfg_not_test.rs:1]: Honestly, no idea, can't remember my reasoning :D

Now, I think it's fine and makes sense

File: clippy_lints/src/cfg_not_test.rs
Comment (Centri3) [inline clippy_lints/src/cfg_not_test.rs:11]: I didn't include it here before because I wasn't entirely sure what it meant when writing this, though this should include the 2nd advantage, like
```rust
/// Enabling this also guards against excessive mockery as well, which is an anti-pattern.
```
Could probably be worded better.

File: tests/ui/cfg_not_test.rs
Comment (Centri3) [reply tests/ui/cfg_not_test.rs:1]: I think we also need a test for `#[cfg(not(not(test))]`. Not sure if that'd lint currently, though it shouldn't. If it does currently, you can track the current negation while visiting the meta items in the `CfgNotTest` struct, and only lint it's negative (an odd number of `not`s).

File: clippy_lints/src/cfg_not_test.rs
Comment (Centri3) [inline clippy_lints/src/cfg_not_test.rs:42]: Major nit, but I think these two should be `_once` diagnostics.

File: clippy_lints/src/cfg_not_test.rs
Comment (MrNossiom) [reply clippy_lints/src/cfg_not_test.rs]: > merging a PR to clean that up a bit.
Would definitely be willing to do that 😉 

> In the case of this lint, it's allow-by-default
I thought restriction was warn-by-default (it's not). In the original issue, @blyxyas states "I feel like the drawback is too heavy for an allow-by-default lint" which I understand as "the lint should be warn-by-default"

File: clippy_lints/src/cfg_not_test.rs
Comment (MrNossiom) [reply clippy_lints/src/cfg_not_test.rs:42]: `help_once` doesn't exist

Comment (Centri3) [reply clippy_lints/src/cfg_not_test.rs:42]: Hmm, thought that was a thing. That's fine then.

File: clippy_lints/src/cfg_not_test.rs
Comment (Centri3) [reply clippy_lints/src/cfg_not_test.rs]: I think she meant enabled-by-default, `restriction` is allow-by-default and generally, lints in that group aren't applicable in every situation or go against idiomatic code. There are valid reasons for using `not(test)`.

**Comment by blyxyas** [comment]
> I thought restriction was warn-by-default (it's not). In the original issue, @blyxyas states "I feel like the drawback is too heavy for an allow-by-default lint" which I understand as "the lint should be warn-by-default"

Totally a typo, it should not be warn by default imo

**Comment by Jarcho** [comment]
LGTM, but since you were already reviewing this. r? @Centri3

**Comment by xFrednet** [comment]
@Centri3 is currently inactive. @Jarcho would you mind taking this review back? If you don't have the time rn, you can reassign it using `r? clippy` or to me :)

r? @Jarcho

**Comment by mrnossiom** [comment]
Hi @Jarcho,

It's been a little time since I've made this PR. I would love to finish it. I remember being blocked by not understanding exactly what was blocking.

If I am correct, early passes are kind of unstable and buggy, but the information I need is not available after `cfg` expansion ? Or are `cfg`'s expanded before early passes ?

**Comment by Jarcho** [comment]
Early passes are fine and have access to the ast before any `cfg` based removal has happened. The passes we don't want to add are pre-expansion passes, but that's not needed for this.

**Comment by mrnossiom** [comment]
So this PR is good as is, then ?

File: clippy_lints/src/cfg_not_test.rs
Comment (Jarcho) [inline clippy_lints/src/cfg_not_test.rs:51]: Should be `is_some_and`. Right now `#[cfg(not(foo))]` will return true.

File: clippy_lints/src/cfg_not_test.rs
Comment (Jarcho) [inline clippy_lints/src/cfg_not_test.rs:53]: You can use `is_some_and` instead of `map_or(false, ...)`.

**Review by Jarcho — CHANGES_REQUESTED** [review]

File: clippy_lints/src/cfg_not_test.rs
Comment (mrnossiom) [reply clippy_lints/src/cfg_not_test.rs]: Should I update tests to match such a case?

Comment (Jarcho) [reply clippy_lints/src/cfg_not_test.rs]: Definitely add a test for it.

**Comment by Jarcho** [comment]
If you haven't already you'll have to rebase and rebless the tests. There was a `ui_test` update that changed the format a bit.

**Comment by xFrednet** [comment]
Hey, this is a ping from triage. @Jarcho can you give this PR a review? It's totally fine if you don't have the time right now, you can reassign the PR to a random team member using `r? clippy`.

@mrnossiom Would you mind rebasing the PR again?

@rustbot ready

**Review by Jarcho — APPROVED** [review]

**Comment by Jarcho** [comment]
Marked this as approved. @rust-lang/clippy can someone merge this if they see it before I do since this is now multiple times I've missed it.

**Pull request merged**
